central/clusterinit/backend/backend.go

```diff
@@ -51,27 +52,10 @@ func newBackend(store store.Store, certProvider certificate.Provider) Backend {
 
 // CheckAccess returns nil if requested access level is granted in context.
 func CheckAccess(ctx context.Context, access storage.Access) error {
-	// we need access to the API token and service identity resources
-	serviceIdentityScopes := [][]sac.ScopeKey{
-		{sac.AccessModeScopeKey(access), sac.ResourceScopeKey(resources.ServiceIdentity.GetResource())},
-		{sac.AccessModeScopeKey(access), sac.ResourceScopeKey(resources.Administration.GetResource())},
-	}
-	apiTokenScopes := [][]sac.ScopeKey{
-		{sac.AccessModeScopeKey(access), sac.ResourceScopeKey(resources.APIToken.GetResource())},
-		{sac.AccessModeScopeKey(access), sac.ResourceScopeKey(resources.Integrations.GetResource())},
-	}
-
-	svcIdentityAllowed, err := sac.GlobalAccessScopeChecker(ctx).AnyAllowed(ctx, serviceIdentityScopes)
-	if err != nil {
-		return errors.Wrap(err, "checking access")
-	}
-	apiTokenAllowed, err := sac.GlobalAccessScopeChecker(ctx).AnyAllowed(ctx, apiTokenScopes)
-	if err != nil {
-		return errors.Wrap(err, "checking access")
-	}
-	if !apiTokenAllowed || !svcIdentityAllowed {
-		return errox.NotAuthorized
-	}
+	authZ := and.And(
+		user.With(permissions.ResourceWithAccess{Resource: resources.ServiceIdentity, Access: access}),
+		user.With(permissions.ResourceWithAccess{Resource: resources.APIToken, Access: access}),
+	)
```

Comment:
I see one disadvantage with this approach right now, maybe I'm also missing the proper interface to use:
If we switch to `authz.Authorizer` instead of `GlobalAccessScopeChecker`, the tests fail due to `CheckRevoked` being called with `sac.WithAllAccess(ctx)`. This context doesn't have proper access to the resources, and it seems to be intended that way. Is there any idiomatic way how we could fix this @rukletsov ?

Reply:
Looks like `clusterinit/backend` doesn't want to deal with identity at that level and prefers `GlobalAccessScopeChecker`. We currently do not have an authorizer implementation that solely relies on the `GlobalAccessScopeChecker` for permissions check, `permissionChecker` does that only in some specific cases. I think it makes sense to use existing `and.And` logic here and also your addition to `evaluateAgainstPermissions()`.

Technically, we can add an authorizer that relies exclusively on GlobalAccessScopeChecker for permissions check and have it share most of the code with `permissionChecker`.

However, I wonder if the universe tries to tell us that we're doing something wrong. Let's have a look at, say, [built-in scoped authorizer](https://github.com/stackrox/stackrox/blob/831b5f2351ddab2f07e69653ed887aadacf0aa28/central/sac/authorizer/builtin_scoped_authorizer.go#L133). It is not aware of the _replacing resource_ concept, and maybe that's fine. But [GlobalAccessScopeChecker](https://github.com/stackrox/stackrox/blob/3b972b936f41178498082a7d0121942333afa1e2/pkg/sac/context.go#L14) is not aware of the concept either so we actually carry a scope checker in our context which is not aware of _replacing resources_. We can mitigate it here but I'm not sure if there are any current or future places that "forget" to do so.

At the moment, I don't know how to extend your approach in `orScopeChecker` to any scope checker or scope checker core. Maybe we shall ensure that every `ScopeCheckerCore` implementation is aware of _replacing resources_? IIRC, there are not that many.

Reply:
Agreed with the sentiment regarding the "universe tries to tell us that we're doing something wrong". The `orScopeChecker` works fine when used with i.e. `sac.ForResourceHelper()` and the `ScopeChecker` interface, but is not solving the replacing resources for `ScopeCheckerCore` implementations, as you mentioned.

I'd propose to go ahead as follows:
- Remove identity from `clusterinit/backend`, resort back to `sac.ForResourceHelpers`, merge this PR.
- Check each `ScopeCheckerCore` implementation and make them aware of replacing resources in a follow-up PR(s).

This way, we can be sure that whatever clients are using, either `ScopeChecker` or `ScopeCheckerCore`, the implementation will be aware of replacing resources. I wonder if we then need `orScopeChecker` at all after all changes.